### PR TITLE
Optimize and unite setting CLOEXEC on fds

### PR DESCRIPTION
**Pull request by kolyshkin**
In case maximum number of open files limit is set too high, both luaext/Pexec() and lib/doScriptExec() spend way too much time trying to set `FD_CLOEXEC` for all those file descriptors that might be open, resulting in severe increase of time it takes to execute rpm or dnf. For example, this happens while running under Docker because:

> $ docker run fedora ulimit -n
> 1048576

One obvious fix is to use procfs to get the actual list of opened file descriptors and iterate over those.

My quick-n-dirty benchmark shows the /proc approach is about 10x faster than iterating through a list of 1024 fds.

Note that the old method is still used in case /proc is not available.

While at it, unite the two implementations, and future(1) proof it by making sure we modify the existing flags.

(1) - currently the only known flag is `FD_CLOEXEC`.

This should fix:
- https://github.com/moby/moby/issues/23137
- https://bugzilla.redhat.com/show_bug.cgi?id=1537564

### CODE REVIEW

**Comment by cgwalters** [comment]
glib code here: https://gitlab.gnome.org/GNOME/glib/blob/487b1fd20c5e494366a82ddc0fa6b53b8bd779ad/glib/gspawn.c#L1207

Also worth looking at what systemd does.

File: luaext/lposix.c [hunk 1]
```diff
@@ -330,26 +330,64 @@ static int Pmkfifo(lua_State *L)		/** mkfifo(path) */
...
+
+	return;
+
+fallback:
+	// iterate over all possible fds
+	for (fd = min_fd; fd < open_max; fd++)
+		set_cloexec(fd);
+}
+
 static int Pexec(lua_State *L)			/** exec(path,[args]) */
```
Comment (Saur2000) [inline luaext/lposix.c]: You need to do the same change for doScriptExec() in lib/rpmscript.c as well. At least in our case, that is where all the time is spent.

File: luaext/lposix.c [hunk 2]
```diff
@@ -330,26 +330,64 @@ static int Pmkfifo(lua_State *L)		/** mkfifo(path) */
 }
 
 
+static void set_cloexec(int fd)
+{
+	int flag = fcntl(fd, F_GETFD);
```
Comment (Saur2000) [inline luaext/lposix.c]: Change "flag" to "flags". Even if FD_CLOEXEC is currently the only defined flag, we may as well make the code forward compatible,

File: luaext/lposix.c [hunk 3]
```diff
@@ -330,26 +330,64 @@ static int Pmkfifo(lua_State *L)		/** mkfifo(path) */
...
+
+	if (flag == -1 || (flag & FD_CLOEXEC))
+		return;
+
+	fcntl(fd, F_SETFD, FD_CLOEXEC);
```
Comment (Saur2000) [inline luaext/lposix.c]: Change "FD_CLOEXEC" to "flags | FD_CLOEXEC".

File: luaext/lposix.c [hunk 4]
```diff
@@ -330,26 +330,64 @@ static int Pmkfifo(lua_State *L)		/** mkfifo(path) */
...
+static void set_cloexec_all(void)
+{
+	const int min_fd = 3; // don't touch stdin/out/err
+	const int default_open_max = 1024;
+	int fd, open_max;
+
+	open_max = sysconf(_SC_OPEN_MAX);
+	if (open_max == -1) {
+	    open_max = default_open_max;
+	    goto fallback;
```
Comment (Saur2000) [inline luaext/lposix.c]: I see no reason to goto fallback here.

File: luaext/lposix.c [hunk 5]
```diff
@@ -330,26 +330,64 @@ static int Pmkfifo(lua_State *L)		/** mkfifo(path) */
...
+	}
+
+	if (open_max <= default_open_max) {
```
Comment (Saur2000) [inline luaext/lposix.c]: I doubt you need to be this conservative. I don't have any measurements to support it, but I'd be very surprised if it is not a win to remove this if statement and always use the code that only traverses the open file descriptors.

Comment (kolyshkin) [reply luaext/lposix.c]: I wanted to be conservative for the sole reason to minimize the change to current behavior. What you say makes sense though, let me update the patch.

**Comment by kolyshkin** [comment]
I did some more benchmarks comparing the new /proc way with the old one. It seems that the old one is faster when `open_max < 60` which should almost never be the case. Based on that and the suggestion from @Saur2000 I'm dropping the conditional switch to old method completely.

**Comment by kolyshkin** [comment]
For the reference, here's my quick-n-dirty benchmarking code:
```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <dirent.h>



static void set_cloexec(int fd)
{
        int flags = fcntl(fd, F_GETFD);

        if (flags == -1 || (flags & FD_CLOEXEC))
                return;

        fcntl(fd, F_SETFD, flags | FD_CLOEXEC);
}

static void set_cloexec_all(int open_max, char method)
{
        const int min_fd = 3; // don't touch stdin/out/err
        int fd;

        if (method == 'i')
		goto fallback;

	// iterate over fds obtained from /proc
        DIR *dir = opendir("/proc/self/fd");
        if (dir == NULL) {
                goto fallback;
        }

        struct dirent *entry;
        while ((entry = readdir(dir)) != NULL) {
                fd = atoi(entry->d_name);
                if (fd >= min_fd)
                        set_cloexec(fd);
        }
        closedir(dir);

        return;

fallback:
        // iterate over all possible fds
        for (fd = min_fd; fd < open_max; fd++)
                set_cloexec(fd);
}

int main(int argc, char **argv) {
	if (argc < 4) {
		fprintf(stderr, "Usage: %s <open_max> <iterations> <method>\n", argv[0]);
		fprintf(stderr, "  <method> is either i (iterate) or p (use proc)\n");
		return 1;
	}

	int open_max = atoi(argv[1]);
	int iter = atoi(argv[2]);
	char m = argv[3][0];
	
	while (iter--)
		set_cloexec_all(open_max, m);

	return 0;
}
```

**Comment by kolyshkin** [comment]
```
kir@kd:~/git/rpm$ time ./a 64 1000000 p

real	0m3.413s
user	0m0.416s
sys	0m2.994s
kir@kd:~/git/rpm$ time ./a 64 1000000 i

real	0m3.661s
user	0m1.421s
sys	0m2.233s
```

**Comment by kolyshkin** [comment]
OK, patch updated to take all the review comments into account (thanks @Saur2000 for your suggestions) so it's now also fixing the `doScriptExec()` from `lib/rpmscript.c`.

**Comment by kolyshkin** [comment]
@cgwalters I took a look at glib implementation, they make use of `fdwalk()` if available (it's a function unique to SunOS/Solaris AFAICS), and reimplement it using /proc/self/fd if not (which is the case for Linux).  This seems to be an unnecessary complication (unless we care much about Solaris).

One other thing is, they call `getrlimit(RLIMIT_NOFILE, &rl)` and use `rl.rlim_max` which seems to be a mistake -- `rlim_cur` should be used. But since this code is only used when /proc is not available it's probably nothing.

**Comment by kolyshkin** [comment]
> Filed https://bugzilla.gnome.org/show_bug.cgi?id=796227

This one is bad, as ulimit might have been changed during the runtime of a process, and so using rlim_max is more correct and safe. For the same reason I have added the second commit https://github.com/rpm-software-management/rpm/pull/444/commits/aa6cc049cc43cc0ae7d7c2d40fe0c9f7d8dcb6f8 to this PR.

**Comment by kolyshkin** [comment]
@pmatilai @ffesti PTAL

**Comment by pmatilai** [comment]
Back from vacation...

This (using /proc/self/fd when available) was what I had in mind, so we're on the right track here, thanks for the work so far. Some quick remarks == requests:
- rpm comment style is "/* foo */", "// foo" should not be used
- just slap the new cloexec function to end of rpmio/rpmio.c (it's quite well at home there, rpmio.c has all/most headers already and this really doesn't need a separate file) and put the prototype to rpmio_internal.h, this is not an interface we want to export
- the jump to fallback label seems klunky and unnecessary, just handle the respective cases in the if-else directly

**Comment by ffesti** [comment]
The patch set would be even nicer if moving the code into its own function would be separated from any changes to the code.

**Comment by kolyshkin** [comment]
@ffesti @pmatilai please see the updated (and rebased) patch set.

> rpm comment style is "/* foo */", "// foo" should not be used

I actually thought about it so I did `git grep '// '` which revealed quite a few uses and so I went ahead.
Anyway, done.

> and put the prototype to rpmio_internal.h

I'm assuming you are OK with `luaext/lposix.c` to have `#include "rpmio/rpmio_internal.h"` added.

> The patch set would be even nicer if moving the code into its own function would be separated from any changes to the code.

done

File: configure.ac [hunk 6]
```diff
@@ -959,7 +959,7 @@ AC_ARG_WITH([lua], [AS_HELP_STRING([--with-lua], [build with lua support])],
 
 AS_IF([test "$with_lua" != no],[
   PKG_CHECK_MODULES([LUA],
-    [lua >= 5.1],
+    [lua53 >= 5.1],
```
Comment (Saur2000) [inline configure.ac]: Was this intentional? It seems unrelated.

File: rpmio/rpmio.c [hunk 7]
```diff
@@ -1760,18 +1760,43 @@ DIGEST_CTX fdDupDigest(FD_t fd, int id)
...
+	struct dirent *entry;
+	while ((entry = readdir(dir)) != NULL) {
+		fd = atoi(entry->d_name);
+		if (fd > min_fd)
+			set_cloexec(fd);
 	}
+
+	closedir(dir);
+
+	return;
```
Comment (Saur2000) [inline rpmio/rpmio.c]: Remove.
